File: samples/java-eventsourced-shopping-cart/src/main/java/com/example/shoppingcart/domain/ShoppingCart.java
/*
 * Copyright 2021 Lightbend Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.shoppingcart.domain;

import kalix.javasdk.eventsourcedentity.CommandContext;
import kalix.javasdk.eventsourcedentity.EventContext;
import kalix.javasdk.eventsourcedentity.EventSourcedEntityContext;
import com.example.shoppingcart.ShoppingCartApi;
import com.example.shoppingcart.domain.ShoppingCartDomain;
import com.google.protobuf.Empty;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Collectors;

// tag::class[]
public class ShoppingCart extends AbstractShoppingCart { // <1>
  @SuppressWarnings("unused")
  private final String entityId;

  public ShoppingCart(EventSourcedEntityContext context) { this.entityId = context.entityId(); }

  @Override
  public ShoppingCartDomain.Cart emptyState() { // <2>
    return ShoppingCartDomain.Cart.getDefaultInstance();
  }
  // end::class[]

  // tag::addItem[]
  @Override
  public Effect<Empty> addItem(
      ShoppingCartDomain.Cart currentState,
      ShoppingCartApi.AddLineItem command) {
    if (currentState.getCheckedOut())
      return effects().error("Cart is already checked out.");
    if (command.getQuantity() <= 0) { // <1>
      return effects().error("Quantity for item " + command.getProductId() + " must be greater than zero.");
    }

    ShoppingCartDomain.ItemAdded event = // <2>
        ShoppingCartDomain.ItemAdded.newBuilder()
            .setItem(
                ShoppingCartDomain.LineItem.newBuilder()
                    .setProductId(command.getProductId())
                    .setName(command.getName())
                    .setQuantity(command.getQuantity())
                    .build())
            .build();

    return effects()
            .emitEvent(event) // <3>
            .thenReply(newState -> Empty.getDefaultInstance()); // <4>
  }
  // end::addItem[]

  @Override
  public Effect<Empty> removeItem(
      ShoppingCartDomain.Cart currentState,
      ShoppingCartApi.RemoveLineItem command) {
    if (currentState.getCheckedOut())
      return effects().error("Cart is already checked out.");
    if (findItemByProductId(currentState, command.getProductId()).isEmpty()) {
      return effects()
          .error(
              "Cannot remove item " + command.getProductId() + " because it is not in the cart.");
    }

    ShoppingCartDomain.ItemRemoved event =
        ShoppingCartDomain.ItemRemoved.newBuilder().setProductId(command.getProductId()).build();

    return effects()
            .emitEvent(event)
            .thenReply(newState -> Empty.getDefaultInstance());
  }

  // tag::getCart[]
  @Override
  public Effect<ShoppingCartApi.Cart> getCart(
      ShoppingCartDomain.Cart currentState, // <1>
      ShoppingCartApi.GetShoppingCart command) {
    List<ShoppingCartApi.LineItem> apiItems =
        currentState.getItemsList().stream()
            .map(this::convert)
            .sorted(Comparator.comparing(ShoppingCartApi.LineItem::getProductId))
            .collect(Collectors.toList());
    ShoppingCartApi.Cart apiCart =
            ShoppingCartApi.Cart.newBuilder().addAllItems(apiItems)
                .setCheckedOut(currentState.getCheckedOut())
                .build(); // <2>
    return effects().reply(apiCart);
  }

  private ShoppingCartApi.LineItem convert(ShoppingCartDomain.LineItem item) {
    return ShoppingCartApi.LineItem.newBuilder()
            .setProductId(item.getProductId())
            .setName(item.getName())
            .setQuantity(item.getQuantity())
            .build();
  }
  // end::getCart[]

  // tag::checkout[]
  @Override
  public Effect<Empty> checkout(ShoppingCartDomain.Cart currentState, ShoppingCartApi.CheckoutShoppingCart checkoutShoppingCart) {
    if (currentState.getCheckedOut())
      return effects().error("Cart is already checked out.");
    return effects()
        .emitEvent(ShoppingCartDomain.CheckedOut.getDefaultInstance()) // <1>
        .deleteEntity() // <2>
        .thenReply(newState -> Empty.getDefaultInstance());
  }
  // end::checkout[]

  // tag::itemAdded[]
  @Override
  public ShoppingCartDomain.Cart itemAdded(
      ShoppingCartDomain.Cart currentState,
      ShoppingCartDomain.ItemAdded itemAdded) {
    ShoppingCartDomain.LineItem item = itemAdded.getItem();
    ShoppingCartDomain.LineItem lineItem = updateItem(item, currentState);
    List<ShoppingCartDomain.LineItem> lineItems =
        removeItemByProductId(currentState, item.getProductId());
    lineItems.add(lineItem);
    lineItems.sort(Comparator.comparing(ShoppingCartDomain.LineItem::getProductId));
    return ShoppingCartDomain.Cart.newBuilder().addAllItems(lineItems).build();
  }

  // end::itemAdded[]

  @Override
  public ShoppingCartDomain.Cart itemRemoved(
      ShoppingCartDomain.Cart currentState,
      ShoppingCartDomain.ItemRemoved itemRemoved) {
    List<ShoppingCartDomain.LineItem> items =
        removeItemByProductId(currentState, itemRemoved.getProductId());
    items.sort(Comparator.comparing(ShoppingCartDomain.LineItem::getProductId));
    return ShoppingCartDomain.Cart.newBuilder().addAllItems(items).build();
  }

  // tag::itemAdded[]
  private ShoppingCartDomain.LineItem updateItem(
      ShoppingCartDomain.LineItem item, ShoppingCartDomain.Cart cart) {
    return findItemByProductId(cart, item.getProductId())
        .map(li -> li.toBuilder().setQuantity(li.getQuantity() + item.getQuantity()).build())
        .orElse(item);
  }

  private Optional<ShoppingCartDomain.LineItem> findItemByProductId(
      ShoppingCartDomain.Cart cart, String productId) {
    Predicate<ShoppingCartDomain.LineItem> lineItemExists =
        lineItem -> lineItem.getProductId().equals(productId);
    return cart.getItemsList().stream().filter(lineItemExists).findFirst();
  }

  private List<ShoppingCartDomain.LineItem> removeItemByProductId(
      ShoppingCartDomain.Cart cart, String productId) {
    return cart.getItemsList().stream()
        .filter(lineItem -> !lineItem.getProductId().equals(productId))
        .collect(Collectors.toList());
  }
  // end::itemAdded[]

  @Override
  public ShoppingCartDomain.Cart checkedOut(
      ShoppingCartDomain.Cart currentState,
      ShoppingCartDomain.CheckedOut checkedOut) {
    return ShoppingCartDomain.Cart.newBuilder(currentState).setCheckedOut(true).build();
  }
}

File: samples/scala-protobuf-eventsourced-shopping-cart/src/main/scala/com/example/shoppingcart/domain/ShoppingCart.scala
package com.example.shoppingcart.domain

// tag::imports[]
import kalix.scalasdk.eventsourcedentity.EventSourcedEntity
import kalix.scalasdk.eventsourcedentity.EventSourcedEntityContext
import com.example.shoppingcart
import com.google.protobuf.empty.Empty
import scala.annotation.nowarn

import com.example.shoppingcart.CheckoutShoppingCart
// end::imports[]

// tag::class[]
class ShoppingCart(context: EventSourcedEntityContext) extends AbstractShoppingCart { // <1>

  @nowarn("msg=unused")
  private val entityId = context.entityId

  override def emptyState: Cart = Cart.defaultInstance // <2>
  // end::class[]

  // tag::addItem[]
  override def addItem(currentState: Cart, addLineItem: shoppingcart.AddLineItem): EventSourcedEntity.Effect[Empty] =
    if (currentState.checkedOut)
      effects.error("Cart is already checked out")
    else if (addLineItem.quantity <= 0)
      effects.error(s"Quantity for item ${addLineItem.productId} must be greater than zero.") // <1>
    else {
      val event = ItemAdded( // <2>
        Some(LineItem(productId = addLineItem.productId, name = addLineItem.name, quantity = addLineItem.quantity)))
      effects
        .emitEvent(event) // <3>
        .thenReply(_ => Empty.defaultInstance) // <4>
    }
  // end::addItem[]

  // tag::removeItem[]
  override def removeItem(
      currentState: Cart,
      removeLineItem: shoppingcart.RemoveLineItem): EventSourcedEntity.Effect[Empty] =
    if (currentState.checkedOut)
      effects.error("Cart is already checked out")
    else if (!currentState.items.exists(_.productId == removeLineItem.productId)) {
      effects.error(s"Cannot remove item ${removeLineItem.productId} because it is not in the cart.")
    } else {
      val event = ItemRemoved(removeLineItem.productId)
      effects
        .emitEvent(event)
        .thenReply(_ => Empty.defaultInstance)
    }
  // end::removeItem[]

  // tag::getCart[]
  override def getCart(
      currentState: Cart, // <1>
      getShoppingCart: shoppingcart.GetShoppingCart): EventSourcedEntity.Effect[shoppingcart.Cart] = {
    val apiItems = currentState.items.map(convertToApi).sortBy(_.productId)
    val apiCart = shoppingcart.Cart(apiItems, currentState.checkedOut) // <2>
    effects.reply(apiCart)
  }

  private def convertToApi(item: LineItem): shoppingcart.LineItem =
    shoppingcart.LineItem(productId = item.productId, name = item.name, quantity = item.quantity)
  // end::getCart[]

  // tag::checkout[]
  override def checkout(
      currentState: Cart,
      checkoutShoppingCart: CheckoutShoppingCart): EventSourcedEntity.Effect[Empty] = {
    if (currentState.checkedOut)
      effects.error("Cart is already checked out")
    else
      effects
        .emitEvent(CheckedOut.defaultInstance) // <1>
        .deleteEntity() // <2>
        .thenReply(_ => Empty.defaultInstance);
  }
  // end::checkout[]

  // tag::itemAdded[]
  override def itemAdded(currentState: Cart, itemAdded: ItemAdded): Cart = {
    val cart = currentState.items.map(lineItem => lineItem.productId -> lineItem).toMap
    val item = cart.get(itemAdded.getItem.productId) match {
      case Some(existing) => existing.copy(quantity = existing.quantity + itemAdded.getItem.quantity)
      case None           => itemAdded.getItem
    }
    val updatedCart = cart + (item.productId -> item)
    currentState.withItems(updatedCart.values.toSeq)
  }
  // end::itemAdded[]

  // tag::itemRemoved[]
  override def itemRemoved(currentState: Cart, itemRemoved: ItemRemoved): Cart =
    currentState.withItems(currentState.items.filterNot(_.productId == itemRemoved.productId))
  // end::itemRemoved[]

  override def checkedOut(currentState: Cart, checkedOut: CheckedOut): Cart =
    currentState.copy(checkedOut = true)
}
